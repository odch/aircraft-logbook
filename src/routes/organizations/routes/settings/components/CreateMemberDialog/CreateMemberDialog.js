import React from 'react'
import PropTypes from 'prop-types'
import { injectIntl, FormattedMessage } from 'react-intl'
import Button from '@material-ui/core/Button'
import Dialog from '@material-ui/core/Dialog'
import DialogActions from '@material-ui/core/DialogActions'
import DialogContent from '@material-ui/core/DialogContent'
import DialogTitle from '@material-ui/core/DialogTitle'
import TextField from '@material-ui/core/TextField'
import CircularProgress from '@material-ui/core/CircularProgress'
import FormControl from '@material-ui/core/FormControl'
import InputLabel from '@material-ui/core/InputLabel'
import Select from '@material-ui/core/Select'
import Input from '@material-ui/core/Input'
import MenuItem from '@material-ui/core/MenuItem'
import Checkbox from '@material-ui/core/Checkbox'
import ListItemText from '@material-ui/core/ListItemText'
import { withStyles } from '@material-ui/core/styles'
import DialogContentText from '@material-ui/core/DialogContentText'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import FormHelperText from '@material-ui/core/FormHelperText'
import { intl as intlShape } from '../../../../../../shapes'

const styles = theme => ({
  loadingIndicator: {
    marginRight: 5
  },
  inviteText: {
    marginTop: theme.spacing(0.2) + 'em'
  }
})

class CreateMemberDialog extends React.Component {
  handleChange = name => e => {
    this.updateData(name, e.target.value)
  }

  handleCheckboxChange = name => e => {
    this.updateData(name, e.target.checked)
  }

  updateData = (name, value) => {
    this.props.updateData({
      [name]: value
    })
  }

  handleClose = () => {
    const { onClose, submitting } = this.props

    if (submitting !== true && onClose) {
      onClose()
    }
  }

  handleSubmit = e => {
    const { onSubmit, organizationId, data, submitting } = this.props

    e.preventDefault()
    if (submitting !== true && onSubmit) {
      onSubmit(organizationId, data)
    }
  }

  msg = id => this.props.intl.formatMessage({ id })

  render() {
    const { roles, submitting, limitReached, classes, onClose } = this.props
    return (
      <Dialog onClose={this.handleClose} data-cy="member-create-dialog" open>
        <DialogTitle>
          <FormattedMessage id="organization.member.create.dialog.title" />
        </DialogTitle>
        <form onSubmit={this.handleSubmit}>
          <DialogContent>
            {this.renderTextField('firstname', true, true)}
            {this.renderTextField('lastname', false, true)}
            {this.renderTextField('nr', false, false)}
            {this.renderMultiSelect('roles', roles)}
            {this.renderCheckbox('instructor')}
            <DialogContentText className={classes.inviteText}>
              {this.msg('organization.member.create.dialog.invitation.text')}
            </DialogContentText>
            <FormControl fullWidth error={limitReached}>
              {this.renderTextField(
                'inviteEmail',
                false,
                false,
                'email',
                limitReached
              )}
              {limitReached && (
                <FormHelperText>
                  <FormattedMessage id="organization.member.create.dialog.invitation.limitreached" />
                </FormHelperText>
              )}
            </FormControl>
          </DialogContent>
          <DialogActions>
            <Button onClick={onClose} disabled={submitting}>
              <FormattedMessage id="organization.member.create.dialog.buttons.cancel" />
            </Button>
            <Button
              type="submit"
              color="primary"
              variant="contained"
              data-cy="create-button"
              disabled={submitting}
            >
              {submitting && (
                <CircularProgress
                  size={16}
                  className={classes.loadingIndicator}
                />
              )}
              <FormattedMessage id="organization.member.create.dialog.buttons.create" />
            </Button>
          </DialogActions>
        </form>
      </Dialog>
    )
  }

  renderTextField(name, autoFocus, required, type = 'text', disabled) {
    return (
      <TextField
        label={this.msg(`organization.member.create.dialog.${name}`)}
        type={type}
        fullWidth
        required={required}
        value={this.props.data[name]}
        onChange={this.handleChange(name)}
        data-cy={`${name}-field`}
        disabled={disabled || this.props.submitting}
        autoFocus={autoFocus}
      />
    )
  }

  renderMultiSelect(name, options) {
    const selectedValues = this.props.data[name] || []
    return (
      <FormControl fullWidth>
        <InputLabel>
          <FormattedMessage id={`organization.member.create.dialog.${name}`} />
        </InputLabel>
        <Select
          value={selectedValues}
          onChange={this.handleChange(name)}
          input={<Input />}
          renderValue={selectedValues =>
            selectedValues
              .map(selected => {
                const selectedOption = options.find(
                  option => option.value === selected
                )
                return selectedOption ? selectedOption.label : selected
              })
              .join(', ')
          }
          disabled={this.props.submitting}
          multiple
        >
          {options.map(option => (
            <MenuItem key={option.value} value={option.value}>
              <Checkbox checked={selectedValues.includes(option.value)} />
              <ListItemText primary={option.label} />
            </MenuItem>
          ))}
        </Select>
      </FormControl>
    )
  }

  renderCheckbox(name) {
    return (
      <FormControlLabel
        value={name}
        control={
          <Checkbox
            color="primary"
            checked={this.props.data[name] === true}
            onChange={this.handleCheckboxChange(name)}
          />
        }
        label={this.msg(`organization.member.create.dialog.${name}`)}
        labelPlacement="end"
        disabled={this.props.submitting}
      />
    )
  }
}

CreateMemberDialog.propTypes = {
  organizationId: PropTypes.string.isRequired,
  data: PropTypes.shape({
    firstname: PropTypes.string,
    lastname: PropTypes.string,
    nr: PropTypes.string,
    inviteEmail: PropTypes.string
  }).isRequired,
  roles: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired
    })
  ).isRequired,
  submitting: PropTypes.bool,
  limitReached: PropTypes.bool,
  onClose: PropTypes.func,
  onSubmit: PropTypes.func,
  updateData: PropTypes.func.isRequired,
  intl: intlShape,
  classes: PropTypes.object.isRequired
}

export default injectIntl(withStyles(styles)(CreateMemberDialog))
