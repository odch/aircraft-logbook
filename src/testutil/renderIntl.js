import React from 'react'
import renderer from 'react-test-renderer'
import { createMount } from '@material-ui/core/test-utils'
import { IntlProvider } from 'react-intl'
import pretty from 'pretty'
import messages from '../messages'

const LOCALE = 'de'

const renderInIntlProvider = component => (
  <IntlProvider locale={LOCALE} messages={messages[LOCALE]}>
    {component}
  </IntlProvider>
)

const renderIntl = (component, options) =>
  renderer.create(renderInIntlProvider(component), options)

export const renderIntlUpdate = (renderer, component, options) =>
  renderer.update(renderInIntlProvider(component), options)

export const renderIntlMaterial = (component, returnPrettyHtml = false) => {
  const rendered = createMount()(renderInIntlProvider(component))
  return returnPrettyHtml ? pretty(rendered.html()) : rendered
}

export default renderIntl
