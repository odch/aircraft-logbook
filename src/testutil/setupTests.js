import * as enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import moment from 'moment'
import momentTimezone from 'moment-timezone'
import { init as initFeatureToggles } from '../util/featureToggles'

enzyme.configure({ adapter: new Adapter() })

moment.locale('de')
momentTimezone.locale('de')

initFeatureToggles()
