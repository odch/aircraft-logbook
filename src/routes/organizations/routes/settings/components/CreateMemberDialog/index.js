import CreateMemberDialog from './CreateMemberDialog'

export default CreateMemberDialog
