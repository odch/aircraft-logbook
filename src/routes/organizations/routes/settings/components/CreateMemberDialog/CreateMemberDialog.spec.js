import React from 'react'
import { renderIntlMaterial } from '../../../../../../testutil/renderIntl'
import CreateMemberDialog from './CreateMemberDialog'

describe('routes', () => {
  describe('organizations', () => {
    describe('routes', () => {
      describe('settings', () => {
        describe('components', () => {
          describe('CreateMemberDialog', () => {
            it('renders correctly', () => {
              const renderedValue = renderIntlMaterial(
                <CreateMemberDialog
                  organizationId="my_org"
                  data={{ firstname: 'Max', lastname: 'Muster' }}
                  updateData={() => {}}
                  open
                />,
                true
              )
              expect(renderedValue).toMatchSnapshot()
            })

            it('renders readonly if submitting', () => {
              const renderedValue = renderIntlMaterial(
                <CreateMemberDialog
                  organizationId="my_org"
                  data={{ firstname: 'Max', lastname: 'Muster' }}
                  updateData={() => {}}
                  submitting
                />,
                true
              )
              expect(renderedValue).toMatchSnapshot()
            })
          })
        })
      })
    })
  })
})
