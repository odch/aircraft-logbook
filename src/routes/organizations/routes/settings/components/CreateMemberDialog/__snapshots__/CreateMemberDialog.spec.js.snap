// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes organizations routes settings components CreateMemberDialog renders correctly 1`] = `
"<div role=\\"presentation\\" class=\\"MuiDialog-root\\" data-cy=\\"member-create-dialog\\" style=\\"position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;\\">
  <div class=\\"MuiBackdrop-root\\" aria-hidden=\\"true\\" style=\\"opacity: 1; -webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;\\"></div>
  <div tabindex=\\"0\\" data-test=\\"sentinelStart\\"></div>
  <div class=\\"MuiDialog-container MuiDialog-scrollPaper\\" style=\\"opacity: 1; -webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;\\" role=\\"document\\" tabindex=\\"-1\\">
    <div class=\\"MuiPaper-root MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiPaper-rounded\\" role=\\"dialog\\">
      <div class=\\"MuiDialogTitle-root\\">
        <h2 class=\\"MuiTypography-root MuiTypography-h6\\">Mitglied hinzufügen</h2>
      </div>
      <form>
        <div class=\\"MuiDialogContent-root\\">
          <div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth\\" data-cy=\\"firstname-field\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled Mui-focused Mui-focused Mui-required Mui-required\\" data-shrink=\\"true\\">Vorname<span class=\\"MuiFormLabel-asterisk MuiInputLabel-asterisk\\"> *</span></label>
            <div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth Mui-focused Mui-focused MuiInputBase-formControl MuiInput-formControl\\"><input aria-invalid=\\"false\\" class=\\"MuiInputBase-input MuiInput-input\\" required=\\"\\" type=\\"text\\" value=\\"Max\\"></div>
          </div>
          <div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth\\" data-cy=\\"lastname-field\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled Mui-required Mui-required\\" data-shrink=\\"true\\">Nachname<span class=\\"MuiFormLabel-asterisk MuiInputLabel-asterisk\\"> *</span></label>
            <div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl\\"><input aria-invalid=\\"false\\" class=\\"MuiInputBase-input MuiInput-input\\" required=\\"\\" type=\\"text\\" value=\\"Muster\\"></div>
          </div>
          <div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth\\" data-cy=\\"nr-field\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated\\" data-shrink=\\"false\\">Mitgliedernummer</label>
            <div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl\\"><input aria-invalid=\\"false\\" class=\\"MuiInputBase-input MuiInput-input\\" type=\\"text\\" value=\\"\\"></div>
          </div>
        </div>
        <div class=\\"MuiDialogActions-root MuiDialogActions-spacing\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\">Abbrechen</span><span class=\\"MuiTouchRipple-root\\"></span></button><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary\\" tabindex=\\"0\\" type=\\"submit\\" data-cy=\\"create-button\\"><span class=\\"MuiButton-label\\">Mitglied hinzufügen</span><span class=\\"MuiTouchRipple-root\\"></span></button></div>
      </form>
    </div>
  </div>
  <div tabindex=\\"0\\" data-test=\\"sentinelEnd\\"></div>
</div>"
`;

exports[`routes organizations routes settings components CreateMemberDialog renders readonly if submitting 1`] = `
"<div role=\\"presentation\\" class=\\"MuiDialog-root\\" data-cy=\\"member-create-dialog\\" style=\\"position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;\\">
  <div class=\\"MuiBackdrop-root\\" aria-hidden=\\"true\\" style=\\"opacity: 1; -webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;\\"></div>
  <div tabindex=\\"0\\" data-test=\\"sentinelStart\\"></div>
  <div class=\\"MuiDialog-container MuiDialog-scrollPaper\\" style=\\"opacity: 1; -webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;\\" role=\\"document\\" tabindex=\\"-1\\">
    <div class=\\"MuiPaper-root MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiPaper-rounded\\" role=\\"dialog\\">
      <div class=\\"MuiDialogTitle-root\\">
        <h2 class=\\"MuiTypography-root MuiTypography-h6\\">Mitglied hinzufügen</h2>
      </div>
      <form>
        <div class=\\"MuiDialogContent-root\\">
          <div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth\\" data-cy=\\"firstname-field\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink Mui-disabled Mui-disabled MuiFormLabel-filled Mui-required Mui-required\\" data-shrink=\\"true\\">Vorname<span class=\\"MuiFormLabel-asterisk MuiInputLabel-asterisk\\"> *</span></label>
            <div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline Mui-disabled Mui-disabled MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl\\"><input aria-invalid=\\"false\\" class=\\"MuiInputBase-input MuiInput-input Mui-disabled Mui-disabled\\" disabled=\\"\\" required=\\"\\" type=\\"text\\" value=\\"Max\\"></div>
          </div>
          <div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth\\" data-cy=\\"lastname-field\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink Mui-disabled Mui-disabled MuiFormLabel-filled Mui-required Mui-required\\" data-shrink=\\"true\\">Nachname<span class=\\"MuiFormLabel-asterisk MuiInputLabel-asterisk\\"> *</span></label>
            <div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline Mui-disabled Mui-disabled MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl\\"><input aria-invalid=\\"false\\" class=\\"MuiInputBase-input MuiInput-input Mui-disabled Mui-disabled\\" disabled=\\"\\" required=\\"\\" type=\\"text\\" value=\\"Muster\\"></div>
          </div>
          <div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth\\" data-cy=\\"nr-field\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-disabled Mui-disabled\\" data-shrink=\\"false\\">Mitgliedernummer</label>
            <div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline Mui-disabled Mui-disabled MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl\\"><input aria-invalid=\\"false\\" class=\\"MuiInputBase-input MuiInput-input Mui-disabled Mui-disabled\\" disabled=\\"\\" type=\\"text\\" value=\\"\\"></div>
          </div>
        </div>
        <div class=\\"MuiDialogActions-root MuiDialogActions-spacing\\"><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text Mui-disabled Mui-disabled\\" tabindex=\\"-1\\" type=\\"button\\" disabled=\\"\\"><span class=\\"MuiButton-label\\">Abbrechen</span></button><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled\\" tabindex=\\"-1\\" type=\\"submit\\" disabled=\\"\\" data-cy=\\"create-button\\"><span class=\\"MuiButton-label\\"><div class=\\"MuiCircularProgress-root CreateMemberDialog-loadingIndicator-1 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate\\" style=\\"width: 16px; height: 16px;\\" role=\\"progressbar\\"><svg class=\\"MuiCircularProgress-svg\\" viewBox=\\"22 22 44 44\\"><circle class=\\"MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate\\" cx=\\"44\\" cy=\\"44\\" r=\\"20.2\\" fill=\\"none\\" stroke-width=\\"3.6\\"></circle></svg></div>Mitglied hinzufügen</span></button></div>
      </form>
    </div>
  </div>
  <div tabindex=\\"0\\" data-test=\\"sentinelEnd\\"></div>
</div>"
`;
